gravitee-apim-gateway/gravitee-apim-gateway-core/src/main/java/io/gravitee/gateway/jupiter/core/v4/endpoint/DefaultEndpointManager.java

```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Override
+    public ManagedEndpoint next() {
+        return next(EndpointCriteria.ENDPOINT_UP);
+    }
+
+    @Override
+    public ManagedEndpoint next(final EndpointCriteria criteria) {
+        final String name = criteria.getName();
+
+        if (name == null) {
+            if (defaultGroup != null && criteria.matches(defaultGroup)) {
+                return defaultGroup.next();
+            }
+        } else {
+            // First try to find an endpoint by name.
+            final ManagedEndpoint managedEndpoint = endpointsByName.get(name);
+
+            if (managedEndpoint != null) {
+                if (criteria.matches(managedEndpoint)) {
+                    return managedEndpoint;
+                }
+                return null;
+            }
+
+            final ManagedEndpointGroup managedGroup = groupsByName.get(name);
+
+            if (managedGroup != null && criteria.matches(managedGroup)) {
+                return managedGroup.next();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        for (EndpointGroup endpointGroup : api.getEndpointGroups()) {
+            final ManagedEndpointGroup managedEndpointGroup = createAndStartGroup(endpointGroup);
+
+            if (defaultGroup == null) {
+                defaultGroup = managedEndpointGroup;
+            }
+        }
+    }
+
+    @Override
+    public DefaultEndpointManager preStop() {
+        for (ManagedEndpoint managedEndpoint : endpointsByName.values()) {
+            try {
+                managedEndpoint.getConnector().preStop();
+            } catch (Exception e) {
+                log.warn("An error occurred when pre-stopping endpoint connector [{}].", managedEndpoint.getDefinition().getName());
+            }
+        }
+
+        return this;
+    }
+
+    @Override
+    public void doStop() throws Exception {
+        for (String endpointName : endpointsByName.keySet()) {
+            removeEndpoint(endpointName);
+        }
+
+        endpointsByName.clear();
+        groupsByName.clear();
+    }
+
+    private ManagedEndpointGroup createAndStartGroup(final EndpointGroup endpointGroup) {
+        final ManagedEndpointGroup managedEndpointGroup = new ManagedEndpointGroup(endpointGroup);
+        groupsByName.put(endpointGroup.getName(), managedEndpointGroup);
+
+        for (Endpoint endpoint : endpointGroup.getEndpoints()) {
+            createAndStartEndpoint(managedEndpointGroup, endpoint);
+        }
+
+        return managedEndpointGroup;
+    }
+
+    private void createAndStartEndpoint(final ManagedEndpointGroup managedEndpointGroup, final Endpoint endpoint) {
+        try {
+            final String configuration = getEndpointConfiguration(managedEndpointGroup.getDefinition(), endpoint);
+            final EndpointConnector connector = endpointConnectorPluginManager
+                .getFactoryById(endpoint.getType())
```

Comment:
What if there is no factory ?

Reply:
It will be ends with an NPE and the endpoint will be ignored.

Reply:
No every thing is covered by the try/catch.

Reply:
I've added a special case to have a proper warn log.